architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+# ADR 002: Finalize specifications for interaction between Validator Set Changes, Punishments and Rewards
+
+## Changelog
+- 23-01-2020: Initial Draft
+
+## Context
+
+Currently, `ValidatorState` is used by punishments and rewards as a source of data to compute their respective slashing
+ratio and rewards amount. Besides this, `ValidatorState` is also used to manage dynamic validator set (i.e.,
+validators can be added or removed based on their network operations).
+
+### Problems with current design
+
+1. Our current design does not delete validator information from `ValidatorState` when a validator either unbonds their
+   funds or gets removed from validator set because of some punishment. The reason for not deleting this information is
+   because it is still required for computing `SlashRatio` while slashing a punished validator. Because of this,
+   validators cannot rejoin validator set (because their information already exists and we cannot add a duplicate
+   entry).
+1. During rewards calculation, we calculate `total_stake` based of the active validator at the time of rewards
```

Comment:
as mentioned in https://github.com/crypto-com/chain/pull/966/files#r372212411 this may be OK -- if I understand it correctly, the only drawback is that in one period, slightly lower amount is emitted, and if one leaves the validator set, they won't get rewarded in other reward period.

Reply:
I think we should take validator set changes into account, both unbonding and deposit, and calculate sum of averages.